source/slang/ir.cpp

```diff
@@ -4536,6 +4562,308 @@ namespace Slang
+        // 2. An `in out` parameter isn't just sugar for a combination
+        // of an `in` and an `out` parameter, and instead represents the
+        // read/write "payload" that was passed in. It should legalize
+        // to a single variable, and we can lower reads/writes of it
+        // directly, rather than introduce an intermediate temporary.
+        //
+        switch( stage )
+        {
+        default:
+            break;
+
+        case Stage::AnyHit:
+        case Stage::Callable:
+        case Stage::ClosestHit:
+        case Stage::Intersection:
+        case Stage::Miss:
+        case Stage::RayGeneration:
+            legalizeRayTracingEntryPointParameterForGLSL(context, pp, paramLayout);
+            return;
+        }
+
+        // Is the parameter type a special pointer type
+        // that indicates the parameter is used for `out`
+        // or `inout` access?
+        if(auto paramPtrType = as<IROutTypeBase>(paramType) )
+        {
+            // Okay, we have the more interesting case here,
+            // where the parameter was being passed by reference.
+            // We are going to create a local variable of the appropriate
+            // type, which will replace the parameter, along with
+            // one or more global variables for the actual input/output.
+
+            auto valueType = paramPtrType->getValueType();
+
+            auto localVariable = builder->emitVar(valueType);
+            auto localVal = ScalarizedVal::address(localVariable);
+
+            if( auto inOutType = as<IRInOutType>(paramPtrType) )
+            {
+                // In the `in out` case we need to declare two
+                // sets of global variables: one for the `in`
+                // side and one for the `out` side.
+                auto globalInputVal = createGLSLGlobalVaryings(
+                    context,
+                    builder, valueType, paramLayout, LayoutResourceKind::VaryingInput, stage);
+
+                assign(builder, localVal, globalInputVal);
+            }
+
+            // Any places where the original parameter was used inside
+            // the function body should instead use the new local variable.
+            // Since the parameter was a pointer, we use the variable instruction
+            // itself (which is an `alloca`d pointer) directly:
+            pp->replaceUsesWith(localVariable);
+
+            // We also need one or more global variabels to write the output to
+            // when the function is done. We create them here.
+            auto globalOutputVal = createGLSLGlobalVaryings(
+                    context,
+                    builder, valueType, paramLayout, LayoutResourceKind::VaryingOutput, stage);
+
+            // Now we need to iterate over all the blocks in the function looking
+            // for any `return*` instructions, so that we can write to the output variable
+            for( auto bb = func->getFirstBlock(); bb; bb = bb->getNextBlock() )
```

Comment:
As before can't I have return insts nested in blocks, conditions etc?

Reply:
I think this is answered by the comments above. The blocks nested in an `IRFunc` are always a flat list, and the control-flow graph is defined via their predecessor/successor relationships (which are implicit in the control-flow edges defined by their terminator instructions).